File: 141-linked-list-cycle/141-linked-list-cycle.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
    dummy = ListNode.new
    dummy.next = head
    slow = dummy
    fast = head
    
    while fast != slow do
        return false if !fast || !fast.next
        slow = slow.next
        fast = fast.next.next
    end
    
    true
end

File: easy/141-linked-list-cycle/141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
const hasCycle = function(head) {
    if (!head) return false;
    
    let slow = head,
        fast = head.next;
    
    while (slow !== fast) {
        if (!fast || !fast.next) return false;
        slow = slow.next;
        fast = fast.next.next;
    }
    
    return true;
};